test/e2e/external_oidc_test.go

```diff
@@ -0,0 +1,461 @@
+package e2e
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	configv1 "github.com/openshift/api/config/v1"
+	configclient "github.com/openshift/client-go/config/clientset/versioned"
+	operatorversionedclient "github.com/openshift/client-go/operator/clientset/versioned"
+	test "github.com/openshift/cluster-authentication-operator/test/library"
+	"github.com/stretchr/testify/require"
+
+	authenticationv1 "k8s.io/api/authentication/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/storage/names"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	// TODO: this duplicates the constant defined here: https://github.com/openshift/api/blob/ce6b9c8ad5e785e337869b3bf0f71deeff7bf0eb/features/features.go#L428
+	// until we resolve a dependency breaking compatibility and are able to reference it directly
+	featureGateExternalOIDC = configv1.FeatureGateName("ExternalOIDC")
+
+	oidcClientId      = "admin-cli"
+	oidcAudience      = "openshift-aud"
+	oidcGroupsClaim   = "groups"
+	oidcUsernameClaim = "email"
+
+	kasNamespace = "openshift-kube-apiserver"
+
+	// set if a specific CA bundle is needed for the OIDC provider
+	oidcCABundleConfigMap          = ""
+	oidcCABundleConfigMapNamespace = ""
+)
+
+type testClient struct {
+	t *testing.T
+
+	kubeConfig           *rest.Config
+	kubeClient           *kubernetes.Clientset
+	configClient         *configclient.Clientset
+	operatorConfigClient *operatorversionedclient.Clientset
+}
+
+type oidcAuthResponse struct {
+	AccessToken      string `json:"access_token"`
+	ExpiresIn        int    `json:"expires_in"`
+	RefreshToken     string `json:"refresh_token"`
+	RefreshExpiresIn int    `json:"refresh_expires_in"`
+	TokenType        string `json:"token_type"`
+	IdToken          string `json:"id_token"`
+	NotBeforePolicy  int    `json:"not_before_policy"`
+	SessionState     string `json:"session_state"`
+	Scope            string `json:"scope"`
+}
+
+func TestExternalOIDCWithKeycloak(t *testing.T) {
+	testCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	tc, err := newTestClient(t)
+	require.NoError(t, err)
+
+	// ===============================
+	// Check ExternalOIDC feature gate
+	// ===============================
+
+	featureGates, err := tc.configClient.ConfigV1().FeatureGates().Get(testCtx, "cluster", metav1.GetOptions{})
+	require.NoError(tc.t, err)
+
+	if len(featureGates.Status.FeatureGates) != 1 {
+		// skip test if there are multiple feature gate versions (i.e. ongoing upgrade)
+		tc.t.Skipf("multiple feature gate versions detected")
```

Comment:
I'd fail instead of skip if the test isn't prepared to handle multiple versions. The risk of skipping is that if this test is later added to an upgrade job config, no one will notice that this functionality is not being tested. Alternatively, check for the feature gate in the latest version in the ClusterVersionStatus?

Reply:
Instead of `Get` featuregate/cluster's `.status.featureGates` array, how about checking below configmap which always reflects the feature gates in `latest version`?
In an env which enables it:
```
$ oc get cm config -n openshift-kube-apiserver -o=jsonpath='{.data.config\.yaml}' \
    | jq '.apiServerArguments."feature-gates"'
...
  "ExternalOIDC=true",
...
```
In an env which does not enable it:
```
$ oc get cm config -n openshift-kube-apiserver -o=jsonpath='{.data.config\.yaml}' \
    | jq '.apiServerArguments."feature-gates"'
...
  "ExternalOIDC=false",
...
```